### ISSUE TITLE
Symbol confusion

### ISSUE THREAD

**Issue by rhc54** [pmix/pmix#1128]
I've had a report from a 3rd party that is attempting to link PMIx, PRRTE, and Portals 4 together in an application. The issue is that PMIx and PRRTE both link against libevent, while Portals 4 links against libev. Sadly, the two event libraries have overlapping symbol names - and are incompatible with each other.

Because all three elements are linked against external versions of the respective event libraries, I don't believe we can use our old symbol renaming methods - I believe those require placing an appropriate header in the build system of the event library to re-define the symbols.

I'm therefore stumped as to how to resolve the problem. Does anyone have a suggestion?

**Comment by rhc54** [comment]
@jsquyres @bwbarrett Adding you guys since you know a lot about this area and may already have solved similar problems.

**Comment by jsquyres** [comment]
I don't know of a good answer here, either.  Mixing multiple libraries that have an overlap of symbols is... no good.

The only way to fix this *might* be to not directly link in libevent (or libev), and instead `dlopen()` libevent (or libev) in a private namespace.  And call everything in that library via function pointer -- and since you *can* control the namespace of the function pointers, you'll likely be ok (e.g., have `pmix_libevent_foo` call the back-end `libevent_foo`).

I think that will work in all cases, even the case where the pmix library is, itself, `dlopen`ed.

But it's a pretty giant disruption.  You have to add all kinds of infrastructure to find and `dlopen()` your target library (thar be dragons thar), and then examine it to make sure it's a compatible version to the header file against which you compiled (even moar dragons).  *Then* you can go ahead and `dlsym()` all the symbols you want and put them in function pointers (no dragons, but lots of repetitive / menial work).

...but before going down that road, let me ask: how has Portals 4 worked with Open MPI for forever?  Even excluding PMIx -- Open MPI has used libevent forever.  Has Portals 4 just recently switched to libev or something?

**Comment by rhc54** [comment]
Not sure - perhaps @bwbarrett knows? Note that it would only show if you build with an external libevent as the internal one has its symbols renamed. It could just be that nobody has tried that combination, or tried it and learned to avoid it.

I'm trying to see if I can make PRRTE able to use either event library. I understand that there are just a couple of functions that cause the problem - might be we don't really need those or can protect them with appropriate configury.

**Comment by jsquyres** [comment]
Mmmm.  Yes, that's an interesting idea: perhaps we could just make a `libev` plugin for our libevent frameworks...?  That might be the easiest path forward.

**Comment by rhc54** [comment]
Well, I was able to make PMIx compile against libev. However, it won't run yet so there is some debugging to do. For anyone interested in poking at it, it is here: https://github.com/pmix/pmix/pull/1129

**Comment by ggouaillardet** [comment]
I just added an inline patch at pmix/pmix#1129 (it fixes stuff, but is not enough to get it working)

a few comments
 - `libev` can do `libevent` emulation, but from what I've read, only the `libevent 1.4.1-beta` API is emulated
 - emulation relies on the same `event.h` header file. On RHEL7, it is provided by both `libevent-devel` and `libev-libevent-devel`. on my box, the former was installed, and I do not think `configure` did detect the issue. `libev` emulation is __not__ ABI compatible to `libevent`
- `EV_NONE` is macro defined in both `elf.h` and is used in an `enum` in `event.h` (I had to `#undef` otherwise I could not build).
- various sources state that `libev` is _better_ than `libevent`, and some even claim the new `libuv` is _even better_.

Bottom line, I am not sure if `libev` emulation is a viable option or not.
Ideally, we would develop our own emulation library so `libevent`, `libev` or even `libuv` can be chosen, but I am also not sure this is an effort we want to invest into.

**Comment by jsquyres** [comment]
@hjelmn Looked at replacing libevent in OMPI with libev for a while.  I think he had promising results, but I don't think there was any "wow, it's soooo much better that we *must* switch ASAP!" kind of compelling data.  Remember that libevent is only used for non-latency-critical stuff (e.g., the TCP BTL).  So minor performance differences between libevent and libev and libuv (I didn't know about that one at all) aren't hugely important to us.

Those who care deeply about the TCP BTL (e.g., AWS) may have better information than me / may care about this more.

**Comment by rhc54** [comment]
This has been fixed by the libev compatibility changes courtesy of @ggouaillardet
